kb/development/script-coding-standards.md

```diff
@@ -0,0 +1,102 @@
+
+## Python coding standards
+
+### Directory structure
+
+Non-executable Python modules are in `scripts/mbedtls_dev`.
+Executable Python scripts in `scripts` can use auxiliary modules with `from mbedtls_dev import …`.
+
+Executable Python scripts in `tests/scripts` cannot refer to modules in `scripts/mbedtls_dev` directly, because we do not assume that `PYTHONPATH` contains the Mbed TLS source tree. To use other Python modules from a script in `tests/scripts`, use the following idiom:
+
+```python
+import scripts_path # pylint: disable=unused-import
+from mbedtls_dev import …
+```
+
+### Python imports
+
+We follow the usual convention that imports go at the top of a file (after the docstring and the license comment). Imports are typically ordered as follows:
+
+1. Standard library, in lexicographic order.
+2. Third-party libraries, if any, in lexicographic order.
+3. Modules from `mbedtls_dev`, in lexicographic order.
+
+Avoid importing individual names from modules, as it can be difficult to figure out which module an unqualified name comes from. It's ok to import names if they're sufficiently recognizable, for example:
+
+* Importing individual modules from `mbedtls_dev` is fine.
+* Importing capitalized type names from `typing` is fine.
+* Importing individual submodules from Cryptodome is fine.
+
+### Third-party Python libraries
+
+Avoid bringing in new dependencies, because each dependency adds risk to the project. Only use dependencies if they have a clear advantage.
+
+The minimum supported version of third-party Python libraries is recorded in `*.requirements.txt` files in the [`scripts` directory](https://github.com/Mbed-TLS/mbedtls/tree/development/scripts). As with the [Python version](#language-versions), keep in mind that long-time support branches may need to support older versions.
+
+### Python language features
+
+The maintainers of Mbed TLS are primarily C programmers, not Python programmers. It's ok to use fancy language features if they're useful, but don't necessarily assume that the reader is familiar with them. Make sure that if a reader is unfamiliar with an advanced feature, they'll be able to easily look it up.
+
+* Classes, methods and simple inheritance can be used freely.
+* Avoid metaclasses and nontrivial dynamic class creation. If you use these features, write documentation that's accessible to non-experts.
+* Do define [special methods](https://docs.python.org/3/reference/datamodel.html#special-method-names) if convenient for comparison, context managers, etc. But avoid customizing attribute access.
+* Use [decorators](https://docs.python.org/3/glossary.html#term-decorator) if they have a clear benefit.
+* Do use [generators](https://docs.python.org/3/glossary.html#term-generator) and [list comprehension](https://docs.python.org/3/glossary.html#term-list-comprehension) where warranted.
+
+### Python typing
+
+Using type annotations is encouraged. We run [mypy](http://mypy-lang.org/) to verify static typing.
```

Comment:
Can these be backported to 2.28 (see above comment about Python versions)?

Reply:
We've been running mypy on the CI since 2.26.

Reply:
I believe there's a minor dragon here to be aware of. To represent a container that contains a specific type, you need `List`, `Tuple` and friends from `typing`:
```python
from typing import List

def f(x: List[str]):
    ...
```
However in Python 3.9, support for type-hinting in this way with builtin container types [was added](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections), so in Python >= 3.9, you can do:
```python
def f(x: list[str]):
    ...
```
I'm not sure if it's worth documenting this separately though, I think it's covered by the general command to "be compatible with Python 3.5".